File: crates/lae-core/src/task_cleanup.rs
//! Tear-down helpers for archive and delete.

use std::collections::HashSet;
use std::path::{Path, PathBuf};

use crate::config::LaeConfig;
use crate::distrobox;
use crate::error::{LaeError, Result};
use crate::hyprland::{self, HyprWindow};
use crate::models::{SessionState, Task};

#[derive(Debug, Clone, serde::Serialize, serde::Deserialize)]
pub struct TaskTeardownPreview {
    pub window_count: usize,
    pub data_dir: PathBuf,
    pub container_name: String,
    pub container_exists: bool,
}

pub fn task_data_dir(config: &LaeConfig, task_id: &str) -> PathBuf {
    config.tasks_base_dir.join(task_id)
}

pub fn preview_teardown(config: &LaeConfig, task: &Task) -> Result<TaskTeardownPreview> {
    Ok(TaskTeardownPreview {
        window_count: count_task_windows(task)?,
        data_dir: task_data_dir(config, task.id.as_str()),
        container_name: task.container_name.clone(),
        container_exists: distrobox::container_exists(&task.container_name),
    })
}

pub fn is_active_task_context(state: &SessionState, task: &Task) -> bool {
    if state.current_task_id.as_deref() == Some(task.id.as_str()) {
        return true;
    }
    if !hyprland::available() {
        return false;
    }
    let Ok(Some(active)) = hyprland::get_active_workspace() else {
        return false;
    };
    let workspace_names: HashSet<String> = task.workspace_names().into_iter().collect();
    workspace_names.contains(&active.name)
}

pub fn count_task_windows(task: &Task) -> Result<usize> {
    Ok(clients_for_task(task)?.len())
}

pub fn clients_for_task(task: &Task) -> Result<Vec<HyprWindow>> {
    if !hyprland::available() {
        return Ok(Vec::new());
    }
    let workspace_names: HashSet<String> = task.workspace_names().into_iter().collect();
    let title_prefix = format!("[{}]", task.id);
    Ok(hyprland::get_clients()?
        .into_iter()
        .filter(|client| {
            workspace_names.contains(&client.workspace_name)
                || client.title.starts_with(&title_prefix)
        })
        .collect())
}

pub fn close_task_windows(task: &Task) -> Result<usize> {
    let clients = clients_for_task(task)?;
    for client in &clients {
        hyprland::close_window(&client.address);
    }
    Ok(clients.len())
}

pub fn stop_task_container(task: &Task) -> Result<()> {
    distrobox::stop_container(&task.container_name)
}

pub fn remove_task_container(task: &Task) -> Result<()> {
    distrobox::remove_container(&task.container_name)
}

pub fn purge_task_windows(state: &mut SessionState, task_id: &str) {
    state.windows.retain(|_, record| {
        record.task_id.as_deref() != Some(task_id)
    });
}

pub fn purge_task_session_keys(state: &mut SessionState, task_id: &str) {
    state.last_workspace.remove(&format!("task:{task_id}"));
    if state.current_task_id.as_deref() == Some(task_id) {
        state.current_task_id = None;
        state.context_mode = crate::models::ContextMode::Default;
    }
}

pub fn remove_task_data_dir(config: &LaeConfig, task: &Task) -> Result<()> {
    let task_home = task_data_dir(config, task.id.as_str());
    if task_home.exists() {
        remove_dir_all(&task_home)?;
    }
    Ok(())
}

fn remove_dir_all(path: &Path) -> Result<()> {
    std::fs::remove_dir_all(path).map_err(|source| LaeError::Write {
        path: path.to_path_buf(),
        source,
    })
}

File: src/tsk/core/task_cleanup.py
"""Task tear-down helpers for archive and delete."""

from __future__ import annotations

import shutil
from pathlib import Path

from tsk.core.models import SessionState, Task
from tsk.integrations import distrobox, hyprland


def task_data_dir(base_dir: Path, task_id: str) -> Path:
    return base_dir / task_id


def is_active_task_context(state: SessionState, task: Task) -> bool:
    if state.current_task_id == task.id:
        return True
    if not hyprland.available():
        return False
    active = hyprland.get_active_workspace()
    if active is None:
        return False
    return active.name in set(task.workspace_names())


def clients_for_task(task: Task) -> list:
    if not hyprland.available():
        return []
    workspace_names = set(task.workspace_names())
    title_prefix = f"[{task.id}]"
    return [
        client
        for client in hyprland.get_clients()
        if client.workspace_name in workspace_names
        or client.title.startswith(title_prefix)
    ]


def close_task_windows(task: Task) -> int:
    clients = clients_for_task(task)
    for client in clients:
        hyprland.close_window(client.address)
    return len(clients)


def stop_task_container(task: Task) -> None:
    distrobox.stop_container(task.container_name)


def remove_task_container(task: Task) -> None:
    distrobox.stop_container(task.container_name)
    distrobox.remove_container(task.container_name)


def purge_task_windows(state: SessionState, task_id: str) -> None:
    stale = [addr for addr, record in state.windows.items() if record.task_id == task_id]
    for addr in stale:
        del state.windows[addr]


def purge_task_session_keys(state: SessionState, task_id: str) -> None:
    state.last_workspace.pop(f"task:{task_id}", None)
    if state.current_task_id == task_id:
        state.current_task_id = None
        from tsk.core.models import ContextMode

        state.context_mode = ContextMode.default


def remove_task_data_dir(base_dir: Path, task: Task) -> None:
    task_home = task_data_dir(base_dir, task.id)
    if task_home.is_dir():
        try:
            shutil.rmtree(task_home)
        except OSError as err:
            import sys

            print(
                f"tsk: delete task {task.id}: remove data dir: {err}",
                file=sys.stderr,
            )
